state-chain/chains/src/sol/api.rs

```diff
@@ -60,35 +71,56 @@ pub trait SolanaEnvironment:
 					SolanaTransactionBuildingError::CannotLookupVaultProgramDataAccount,
 				SolanaEnvAccountLookupKey::UpgradeManagerProgramDataAccount =>
 					SolanaTransactionBuildingError::CannotLookupUpgradeManagerProgramDataAccount,
-				SolanaEnvAccountLookupKey::TokenMintPubkey =>
-					SolanaTransactionBuildingError::CannotLookupTokenMintPubkey,
-				SolanaEnvAccountLookupKey::TokenVaultAssociatedTokenAccount =>
-					SolanaTransactionBuildingError::CannotLookupTokenVaultAssociatedTokenAccount,
-				SolanaEnvAccountLookupKey::TokenVaultPdaAccount =>
-					SolanaTransactionBuildingError::CannotLookupTokenVaultPdaAccount,
 			},
 		)
 	}
 
+	fn lookup_account_retain(
+		key: SolanaEnvAccountLookupKey,
+		sol_environment: &mut HashMap<SolanaEnvAccountLookupKey, SolAddress>,
+	) -> Result<SolAddress, SolanaTransactionBuildingError> {
+		match sol_environment.get(&key) {
+			Some(address) => Ok(*address),
+			None => Self::lookup_account(key).map(|addr| {
+				sol_environment.insert(key, addr);
+				addr
+			}),
+		}
+	}
+
 	fn all_nonce_accounts() -> Result<Vec<SolAddress>, SolanaTransactionBuildingError> {
 		<Self as ChainEnvironment<AllNonceAccounts, Vec<(SolAddress, SolHash)>>>::lookup(
 			AllNonceAccounts,
 		)
 		.map(|nonces| nonces.into_iter().map(|(addr, _hash)| addr).collect::<Vec<_>>())
 		.ok_or(SolanaTransactionBuildingError::NoNonceAccountsSet)
 	}
+
+	fn get_token_environment(
+		token_environments: &mut HashMap<SolAsset, TokenEnvironment>,
+		asset: SolAsset,
```

Comment:
Same here, I would suggest either abstracting away the cache implementation behind the call to `lookup`, or just removing the cache. Seems like premature opotimisation.